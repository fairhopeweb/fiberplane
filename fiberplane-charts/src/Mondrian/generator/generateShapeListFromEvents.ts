import type { ProviderEvent } from "../../providerTypes";
import type { Axis, SeriesSource, ShapeList } from "../types";
import { getTimeFromTimestamp } from "./utils";

export function generateShapeListFromEvents(
  { minValue, maxValue }: Axis,
  events: Array<ProviderEvent>,
): ShapeList<SeriesSource, ProviderEvent> {
  return {
    shapes: events.map((event) => {
      const x =
        (getTimeFromTimestamp(event.time) - minValue) / (maxValue - minValue);
      return {
        type: "area",
        points: [{ x, yMin: 0, yMax: 1, source: event }],
        strokeDasharray: [2],
      };
    }),
    source: { type: "events" },
  };
}
