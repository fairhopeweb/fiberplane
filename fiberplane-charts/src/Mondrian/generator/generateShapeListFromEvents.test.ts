import type { ProviderEvent } from "../../providerTypes";
import { generateShapeListFromEvents } from "./generateShapeListFromEvents";

test("it generates shapes from events", () => {
  const events: Array<ProviderEvent> = [
    getEventAtMinute(2, "deploy 1"),
    getEventAtMinute(7, "deploy 2"),
  ];

  expect(
    generateShapeListFromEvents(
      {
        minValue: getDateAtMinute(0).getTime(),
        maxValue: getDateAtMinute(10).getTime(),
      },
      events,
    ),
  ).toMatchSnapshot();
});

function getEventAtMinute(minute: number, title: string): ProviderEvent {
  return {
    time: getDateAtMinute(minute).toISOString(),
    title,
    labels: {},
    attributes: {},
    resource: {},
  };
}

function getDateAtMinute(minute: number): Date {
  return new Date(
    `2023-07-18T16:${minute.toString().padStart(2, "0")}:00.000Z`,
  );
}
