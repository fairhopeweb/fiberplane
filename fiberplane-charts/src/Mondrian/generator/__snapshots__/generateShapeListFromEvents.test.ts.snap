// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it generates shapes from events 1`] = `
{
  "shapes": [
    {
      "points": [
        {
          "source": {
            "attributes": {},
            "labels": {},
            "resource": {},
            "time": "2023-07-18T16:02:00.000Z",
            "title": "deploy 1",
          },
          "x": 0.2,
          "yMax": 1,
          "yMin": 0,
        },
      ],
      "type": "area",
    },
    {
      "points": [
        {
          "source": {
            "attributes": {},
            "labels": {},
            "resource": {},
            "time": "2023-07-18T16:07:00.000Z",
            "title": "deploy 2",
          },
          "x": 0.7,
          "yMax": 1,
          "yMin": 0,
        },
      ],
      "type": "area",
    },
  ],
  "source": {
    "type": "events",
  },
}
`;
